src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/ProcessingSystemCompositeCookers.cs

```diff
@@ -0,0 +1,240 @@
+                // The cooker already exists and is ready to use.
+                return dataCooker;
+            }
+
+            var compositeCookerReference = this.dataExtensionRepository.GetCompositeDataCookerReference(cookerPath);
+            if (compositeCookerReference == null)
+            {
+                throw new ArgumentException(
+                    $"The data extension repository is missing expected composite data cooker: {cookerPath}");
+            }
+
+            if (compositeCookerReference.Availability != DataExtensionAvailability.Available)
+            {
+                throw new ArgumentException(
+                    $"The data cooker is not available: {cookerPath}");
+            }
+
+            var compositeCookerRequiredData = createDataRetrieval(cookerPath);
+            if (compositeCookerRequiredData == null)
+            {
+                // is this a bug in the runtime?
+                Debug.Assert(false);
+                throw new InvalidOperationException(
+                    $"Failed to create {nameof(IDataExtensionRetrieval)} needed for the composite cooker.");
+            }
+
+            var compositeCooker = compositeCookerReference.CreateInstance(compositeCookerRequiredData);
+            if (compositeCooker == null)
+            {
+                Debug.Assert(false, "If the data cooker reference is available, this shouldn't fail.");
+                throw new InvalidOperationException(
+                    $"The composite cooker reference returned null cooker: {cookerPath}. Was it properly initialized?");
+            }
+
+            return this.compositeCookersByPath.GetOrAdd(cookerPath, compositeCooker);
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void ThrowIfDisposed()
+        {
+            if (this.disposedValue)
+            {
+                throw new ObjectDisposedException(this.GetType().Name);
+            }
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (!this.disposedValue)
+            {
+                if (disposing)
+                {
+                    foreach (var cooker in this.compositeCookersByPath.Values)
+                    {
+                        cooker.TryDispose();
+                    }
+
+                    this.compositeCookersByPath.Clear();
+                }
+
+                this.dataExtensionRepository = null;
+                this.retrievalFactory = null;
+                this.disposedValue = true;
+            }
+        }
+
+        /// <summary>
+        ///     This class uses <see cref="ProcessingSystemCompositeCookers"/> to create and store composite
+        ///     cookers, but uses a its own <see cref="IDataExtensionRetrievalFactory"/> which has access
```

Comment:
uses its own factory for what?

Reply:
I think the comments on IDataExtensionRetrievalFactory should answer this question. We can chat if you disagree.